### DISCUSSION TITLE
Mesh UVs

### DISCUSSION THREAD

**Discussion by zalo** [elalish/manifold#683] [Ideas]
It would be nice if MeshGLs came out with UVs.

I added [UV Baking to my CAD Library a little while ago](https://github.com/zalo/CascadeStudio/pull/83), and it ended up being a great decision for the general utility of the outputs.

Here's what Manifold's UVs look like when baked with the [xatlas library](https://github.com/jpcy/xatlas):
![image](https://github.com/elalish/manifold/assets/174475/93183dae-1ca5-42e0-a830-e5c879207499)

[xatlas](https://github.com/jpcy/xatlas) is a small MIT-Licensed dependency (only a single .h and .cpp file) which computed the above unwrap in 56ms.

However, a more ambitious option would be to preserve Mesh UVs across CSG Operations.  This would mean that the UV Space of each of the contributing meshes would have to repack themselves into a larger shared atlas (where the `Manifold` object would have a rect representation of the contributing UVs spaces going into to enable this atlas packing at export time).

It would also mean that all of the contributing primitives would need to come with default UVs, but that's relatively simple.

**Comment by elalish** [comment]
So, UV mapping has a million tradeoffs and I don't want to pick one way and pretend it beats all the other optimizations one could do. I do have the general concept of properties, which are designed to handle UVs, among other things. It is actually common in graphics that the input meshes have separate materials and hence independent UV maps - atlasing is often helpful, but often not. I definitely don't want us to be opinionated about it. 

That said, having the option to generate a UV map sounds very useful indeed. Probably a API akin to `CalculateCurvature`. This sounds like a good library to depend on.
